### PR TITLE
storage: Implement Direr tests

### PR DESCRIPTION
**Pull request by JinnyYi**
ref: https://github.com/beyondstorage/go-storage/issues/560

### CODE REVIEW

File: direr.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+func TestDirer(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", t, func() {
+		Convey("The Storager should implement Direr", func() {
+			_, ok := store.(types.Direr)
+			So(ok, ShouldBeTrue)
+		})
+
+		Convey("When CreateDir", func() {
```
Comment (Xuanwo) [inline direr.go:20]: As we discussed in https://github.com/beyondstorage/go-storage/issues/560#issuecomment-853751578, do we need to reflect them in the tests?

Comment (JinnyYi) [reply direr.go:20]: Should we add the following cases:
- `CreateDir` with `path` not end with `/`
- `CreateDir` with `path` end with `/`
- `CreateDir` with `workDir`
- `List` an empty dir
- `List` a dir within files
The results of creating an existing dir and file and directory with the same name are different between services, so we needn't to add the related cases?

Comment (Xuanwo) [reply direr.go:20]: > CreateDir with path not end with /

LGTM

> CreateDir with path end with /

Service should return error, but the error should be defined.

> CreateDir with workDir

Maybe we should ignore `dir exists` related error here?

> List an empty dir
> List a dir within files

I think they are not related.

---

Here comes up a new problem, for object storage services: If we Stat(“abc”) after CreateDir("abc"), we will get a not exist. We need to find a way to figure it out.

Comment (JinnyYi) [reply direr.go:19]: >If we Stat(“abc”) after CreateDir("abc"), we will get a not exist. We need to find a way to figure it out.
Return not exist when neither `abc` and `abc/` exist？
Same problem with `Delete`, we don't kown whether to delete the file or dir.

Comment (JinnyYi) [reply direr.go:19]: >If we Stat(“abc”) after CreateDir("abc"), we will get a not exist. We need to find a way to figure it out.

Return not exist when neither abc and abc/ exist？
Same problem with Delete, we don't kown whether to delete the file or dir.

Comment (JinnyYi) [reply direr.go:20]: >If we Stat(“abc”) after CreateDir("abc"), we will get a not exist. We need to find a way to figure it out.

Return not exist when neither `abc` nor `abc/` exist？
Same problem with `Delete`, we don't kown whether to delete the file or dir.

Comment (Xuanwo) [reply direr.go:20]: Yes, it's indeed a problem.

Comment (Xuanwo) [reply direr.go:20]: Could be resolved via https://github.com/beyondstorage/specs/pull/93

File: direr.go [hunk 2]
```diff
@@ -16,19 +19,81 @@ func TestDirer(t *testing.T, store types.Storager) {
...
+				So(err, ShouldBeNil)
+			})
+
+			Convey("The Object Mode should be dir", func() {
+				// Dir object's mode must be Dir.
+				So(o.Mode.IsDir(), ShouldBeTrue)
+			})
+		})
+
+		Convey("When CreateDir with path end with `/`", func() {
```
Comment (Xuanwo) [inline direr.go]: For now, we will not check the error when the user inputs `test/`, just ignore it.

File: direr.go [hunk 3]
```diff
@@ -1,11 +1,14 @@
 package tests
 
 import (
+	"errors"
+	"github.com/beyondstorage/go-storage/v4/services"
```
Comment (Xuanwo) [inline direr.go]: Let's format the import.

File: direr.go [hunk 4]
```diff
@@ -16,19 +19,81 @@ func TestDirer(t *testing.T, store types.Storager) {
 			So(ok, ShouldBeTrue)
 		})
 
-		Convey("When CreateDir", func() {
+		Convey("When CreateDir with path not end with `/`", func() {
```
Comment (Xuanwo) [inline direr.go]: `When CreateDir` is enough.

File: direr.go (hunk 2 above)
Comment (JinnyYi) [reply direr.go]: Does it mean the `path` is valid for services and the path check will be added after https://github.com/beyondstorage/specs/issues/95 implemented? Also we don't need to check the `path` in services?

Comment (Xuanwo) [reply direr.go]: Yes.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**